### PR TITLE
feat(swiper): blurhash support for swiper

### PR DESCRIPTION
**Pull request by ferferga**
* Refactored BlurhashImage so it works with any type of image
* Added blurhash to backdrops and logos in Swiper.

gif

### CODE REVIEW

File: components/BlurhashImage.vue
Comment (ferferga) [inline components/BlurhashImage.vue:80]: I know we shouldn't use console, but throw halts the client altogether. I don't think silencing them is a good idea either, so what do you think?

**Comment by MrTimscampi** [comment]
I think logos shouldn't have blurhashes, since they always have transparency (Which is known to cause issues with blurhash, since that doesn't support transparency)

**Comment by ferferga** [comment]
@MrTimscampi API-wise they do

File: components/HomeHeader.vue
Comment (MrTimscampi) [inline components/HomeHeader.vue:18]: Mentioned in my logo PR, but logos shouldn't use blurhash, since they have transparency.

**Comment by MrTimscampi** [comment]
> @MrTimscampi API-wise they do

They do in the API, but that's not a reason to use them :) Transparency with blurhash looks terrible, since it just slaps a white background on things. It's nice for fully opaque images, but we should avoid it in cases where there is transparency, imo.

**Comment by ferferga** [comment]
@MrTimscampi Any specific tag we should ignore aside from Logos so I can add them to my commit?

**Comment by Camc314** [comment]
The only other possible ones are `Art` and `Disk`. I'm not sure where those are used or what they look like so I may be wrong.

File: components/BlurhashImage.vue
Comment (Camc314) [inline components/BlurhashImage.vue:78]: This looks very similar to #306. In it the following is addressed:

> Changes
> - Show movie backdrop instead of primary image when in resume home section
> - Show episode primary image when in resume/up next home section
> - Fall back to backdrop if there is no thumbnail

Comment (ferferga) [reply components/BlurhashImage.vue:78]: Worked on this before knowing anything about #306 

I haven't explained it here so much (mostly in Matrix) but I want to make this component as modular as possible with image types so, whenever we have any item object, we can pass it with an image type confidently, knowing that it will handle all the work for us. We rely too much on the mixin right now (which is still great to have in some cases), but for some basic stuff we can simplify our lives a lot by using a single component.

We can simply create an array of image types that we want to ignore, so we skip the canvas altogether and just keep the image for cases where we pass an image type where we don't want blurhashes.

In a future PR, I'll wrap most images in a LazyImage component that depends on BlurhashImage, so we can have the unloading effect we have in current web without messing too much (see #275). BlurhashImage will be the center of all of it.

Would be great to have all your input on this. cc @MrTimscampi

**Comment by MrTimscampi** [comment]
Trying to test this, getting the following:
image

**Comment by ferferga** [comment]
@MrTimscampi #382 fixes this

File: components/BlurhashImage.vue
Comment (ThibaultNocchi) [reply components/BlurhashImage.vue:80]: I've let a few of console.error where it is critical, but @MrTimscampi wanted to look more into it for modal stacktraces. So I think for now console errors are good. You can disable the linter going crazy with `// eslint-disable-next-line no-console` on the line before.

**Comment by ferferga** [comment]
Blocked until #382 is merged

File: components/HomeHeader.vue
Comment (MrTimscampi) [inline components/HomeHeader.vue:158]: We should need this, since we should have everything required in what's returned by the latest endpoint (If we don't, it's better to add the info there (for things like blurhashes and such), imo.

File: components/HomeHeader.vue
Comment (MrTimscampi) [inline components/HomeHeader.vue:15]:
```suggestion
        <blurhash-image
          v-if="getRelatedItem(item)"
          :key="`${item.Id}-${reloadSentinel}`"
          :item="getRelatedItem(item)"
          :type="'Backdrop'"
        />
```

This fixes my `this.item cannot be undefined` error, but then some items don't have a backdrop at all:
image

File: components/HomeHeader.vue
Comment (ferferga) [reply components/HomeHeader.vue:158]: @mrtimscampi We don't: what we do here is to get parent items.

This function runs the exact same checks you added for getting the image backdrops: for music, it fetched artist, for series, it fetched series item.

As we now use BlurhashImage, we need to pass the full item object. That's where this function enters the game: it requests the full item object of the items that use the backdrop from the parent.

I store those "Related Items" in a dictionary (key is the index of the element in the this.items array) and value this related item.

File: components/HomeHeader.vue
Comment (ferferga) [reply components/HomeHeader.vue:15]: @Mrtimscampi Now I see the problem.

I only return items that have related items in the function, so items that must display its own background images are completely skipped.

I'm going to push a fix, but it doesn't fully solves the issue, as we don't do any filtering in the initial items to check that everything will display a photo, so items without metadata will still be displayed totally black.

Comment (MrTimscampi) [reply components/HomeHeader.vue:15]: >  but it doesn't fully solves the issue, as we don't do any filtering in the initial items to check that everything will display a photo

That's not an issue :) We should show the latest, no matter if it has a backdrop or not. But these had a backdrop and worked before, hence the bug :yum:

File: components/BlurhashImage.vue
Comment (MrTimscampi) [inline components/BlurhashImage.vue:136]: We shouldn't output a message for every error. Some, like this one, can just deal with an even sent back up.

File: components/BlurhashImage.vue
Comment (MrTimscampi) [inline components/BlurhashImage.vue:93]: This doesn't check if the Blurhash is actually valid, just if it exists in the item.

We should perform a check via Blurhash's `isBlurhashValid` function, too.

File: components/HomeHeader.vue
Comment (MrTimscampi) [inline components/HomeHeader.vue:3]:
```suggestion
    v-if="items.length > 0 && !$vuetify.breakpoint.mobile && !loading"
```

We've been using `loading` everywhere else, it's probably best to keep this consistent.

File: components/HomeHeader.vue
Comment (MrTimscampi) [reply components/HomeHeader.vue:158]: > As we now use BlurhashImage, we need to pass the full item object. That's where this function enters the game: it requests the full item object of the items that use the backdrop from the parent.

I still think we should change the server to include the blurhashes from the parent items as well.

We already get the image tags and item IDs for everything, so it'd make sense to get that as well, to minimize requests.

However, for now, it's necessary. Please add a TODO or HACK mentioning we need to go back to this when the server changes have been made, though.

File: components/HomeHeader.vue
Comment (MrTimscampi) [inline components/HomeHeader.vue:185]:
```suggestion
    onSlideChange(swiperInstance: Swiper): void {
```

You already get the Swiper instance on all events emitted by Swiper itself, and the Vue library only passes them through. See https://swiperjs.com/api/#events

File: components/HomeHeader.vue
Comment (MrTimscampi) [inline components/HomeHeader.vue:192]:
```suggestion
    // HACK: Vue-awesome-swiper seems to have a bug where the components inside of duplicated slides (when loop is enabled,
```

Comment (MrTimscampi) [reply components/HomeHeader.vue:192]: We should also probably report this upstream: https://github.com/surmon-china/vue-awesome-swiper

**Review by MrTimscampi — CHANGES_REQUESTED** [review]

File: components/BlurhashImage.vue
Comment (ferferga) [reply components/BlurhashImage.vue:93]: @MrTimscampi That's actually done in the worker with a try/catch (the function performs a halfway decode and it reported false positives for me, so I think doing the decode directly and catching when needing is our best bet)

Comment (MrTimscampi) [reply components/BlurhashImage.vue:93]: > it reported false positives

This might be a bug in our implementation server-side. I know we at least have a color channel issue, since results are pretty different between our custom C# library and Blurhash's reference one in terms of color.

But that sounds good for now.

File: components/HomeHeader.vue
Comment (ferferga) [reply components/HomeHeader.vue:192]: Sure, will do once this is merged (I think it's better than pointing the original dev to a branch that will be deleted after merging)

File: components/HomeHeader.vue
Comment (ferferga) [reply components/HomeHeader.vue]: @MrTimscampi ``swiperInstance`` returns ``undefined``

**Review by MrTimscampi — APPROVED** [review]
LGTM :+1:

**Pull request merged**
